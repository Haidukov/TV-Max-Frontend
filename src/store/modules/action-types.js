export * from './shows/action-types';
export * from './episodes/action-types';
