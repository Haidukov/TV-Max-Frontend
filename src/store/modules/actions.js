export * from './shows/actions';
export * from './episodes/actions';
