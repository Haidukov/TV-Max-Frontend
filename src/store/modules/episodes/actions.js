import {
  FETCH_EPISODES_ERROR,
  FETCH_EPISODES_REQUEST,
  FETCH_EPISODES_SUCCESS,
  FETCH_EPISODE_ERROR,
  FETCH_EPISODE_REQUEST,
  FETCH_EPISODE_SUCCESS
} from '../action-types';

export const fetchEpisodesRequest = id => ({
  type: FETCH_EPISODES_REQUEST,
  payload: { id }
});

export const fetchEpisodesSuccess = (id, episodes) => ({
  type: FETCH_EPISODES_SUCCESS,
  payload: {
    id,
    episodes
  }
});

export const fetchEpisodesError = (id, error) => ({
  type: FETCH_EPISODES_ERROR,
  payload: {
    id,
    error
  }
});

export const fetchEpisodeRequest = id => ({
  type: FETCH_EPISODE_REQUEST,
  payload: { id }
});

export const fetchEpisodeSuccess = (id, data) => ({
  type: FETCH_EPISODE_SUCCESS,
  payload: {
    id,
    data
  }
});

export const fetchEpisodeError = (id, error) => ({
  type: FETCH_EPISODE_ERROR,
  payload: {
    id,
    error
  }
});
