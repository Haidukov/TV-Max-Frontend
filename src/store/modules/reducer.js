import { combineReducers } from 'redux';
import showsReducer from './shows/reducer';
import episodesReducer from './episodes/reducer';

export default combineReducers({
  shows: showsReducer,
  episodes: episodesReducer
});
