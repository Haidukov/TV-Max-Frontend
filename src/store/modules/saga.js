import { all } from 'redux-saga/effects';
import showsSaga from './shows/saga';
import episodesSaga from './episodes/saga';

export default function* rootSaga() {
  yield all([
    showsSaga(),
    episodesSaga()
  ]);
}
