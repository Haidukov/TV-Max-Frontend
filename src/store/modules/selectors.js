export * from './shows/selectors';
export * from './episodes/selectors';
