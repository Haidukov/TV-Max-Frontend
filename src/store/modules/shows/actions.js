import {
  FETCH_SHOW_ERROR,
  FETCH_SHOW_REQUEST,
  FETCH_SHOW_SUCCESS
} from '../action-types';

export const fetchShowRequest = id => ({
  type: FETCH_SHOW_REQUEST,
  payload: { id }
});

export const fetchShowSuccess = (id, data) => ({
  type: FETCH_SHOW_SUCCESS,
  payload: {
    id,
    data
  }
});

export const fetchShowError = (id, error) => ({
  type: FETCH_SHOW_ERROR,
  payload: {
    id,
    error
  }
});
