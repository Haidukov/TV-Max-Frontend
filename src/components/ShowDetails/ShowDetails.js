/* eslint-disable react/no-danger */
import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useParams } from 'react-router';
import { fetchShowRequest } from '../../store/modules/actions';
import {
  getShow,
  getShowError,
  getShowLoading
} from '../../store/modules/selectors';
import DetailsPage from '../Shared/DetailsPage/DetailsPage';
import ErrorMessage from '../Shared/ErrorMessage';
import PageSpinner from '../Shared/PageSpinner';

const ShowDetails = () => {
  const { showId } = useParams();
  const dispatch = useDispatch();

  const show = useSelector(getShow(showId));
  const isLoading = useSelector(getShowLoading(showId));
  const error = useSelector(getShowError(showId));

  useEffect(() => {
    dispatch(fetchShowRequest(showId));
  }, []);

  if (isLoading) {
    return <PageSpinner />;
  }

  if (error) {
    return <ErrorMessage>Failed to load show data</ErrorMessage>;
  }

  return (
    <DetailsPage
      name={show.name}
      imageUrl={show.image?.medium}
      summary={show.summary}
    />
  );
};

export default ShowDetails;
