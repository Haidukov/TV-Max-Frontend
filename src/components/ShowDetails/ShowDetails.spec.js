import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter, Route } from 'react-router';
import { render } from '@testing-library/react';
import ShowDetails from './ShowDetails';
import configureStore from '../../store';
import { fetchShowRequest } from '../../store/modules/actions';
import show from '../../fixtures/show.json';

const renderComponent = (id, partialState = {}) => {
  const preloadedState = {
    shows: {
      byId: {
        [id]: partialState
      }
    }
  };
  const store = configureStore(preloadedState);
  jest.spyOn(store, 'dispatch').mockImplementation(() => {});
  const wrapper = render(
    <MemoryRouter initialEntries={[`/path/${id}`]}>
      <Provider store={store}>
        <Route path='/path/:showId' component={ShowDetails} />
      </Provider>
    </MemoryRouter>
  );
  return { store, wrapper };
};

const findSpinner = wrapper => wrapper.queryByRole('status');

const findErrorMessage = wrapper => wrapper.queryByRole('alert');

const findDetails = wrapper => wrapper.queryByTestId('details');

describe('<ShowDetails />', () => {
  it('should fetch show', () => {
    const { store } = renderComponent(show.id);
    expect(store.dispatch).toHaveBeenCalledWith(fetchShowRequest(show.id.toString()));
  });

  describe('When loading show', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = renderComponent(show.id, { data: null, isLoading: true }).wrapper;
    });

    it('should display loading indicator', () => {
      expect(findSpinner(wrapper)).toBeInTheDocument();
    });

    it('shouldn`t display error message', () => {
      expect(findErrorMessage(wrapper)).not.toBeInTheDocument();
    });

    it('shouldn`t display details', () => {
      expect(findDetails(wrapper)).not.toBeInTheDocument();
    });

    it('should match snapshot', () => {
      expect(wrapper.asFragment()).toMatchSnapshot();
    });
  });

  describe('When error occured', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = renderComponent(show.id, {
        data: null,
        isLoading: false,
        error: new Error('error')
      }).wrapper;
    });

    it('shouldn`t display loading indicator', () => {
      expect(findSpinner(wrapper)).not.toBeInTheDocument();
    });

    it('should display error message', () => {
      expect(findErrorMessage(wrapper)).toBeInTheDocument();
    });

    it('shouldn`t display details', () => {
      expect(findDetails(wrapper)).not.toBeInTheDocument();
    });

    it('should match snapshot', () => {
      expect(wrapper.asFragment()).toMatchSnapshot();
    });
  });

  describe('When data loaded', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = renderComponent(show.id, {
        data: show,
        isLoading: false,
        error: null
      }).wrapper;
    });

    it('shouldn`t display loading indicator', () => {
      expect(findSpinner(wrapper)).not.toBeInTheDocument();
    });

    it('shouldn`t display error message', () => {
      expect(findErrorMessage(wrapper)).not.toBeInTheDocument();
    });

    it('should display details', () => {
      expect(findDetails(wrapper)).toBeInTheDocument();
    });

    it('should match snapshot', () => {
      expect(wrapper.asFragment()).toMatchSnapshot();
    });
  });
});
