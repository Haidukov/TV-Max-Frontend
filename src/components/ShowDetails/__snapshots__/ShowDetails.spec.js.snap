// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ShowDetails /> When data loaded should match snapshot 1`] = `
<DocumentFragment>
  <main
    data-testid="details"
  >
    <h1>
      Person of Interest
    </h1>
    <div
      class="details"
    >
      <img
        alt="Person of Interest"
        class="logo"
        src="https://static.tvmaze.com/uploads/images/medium_portrait/163/407679.jpg"
      />
      <div>
        <p>
          You are being watched. The government has a secret system, a machine that spies on you every hour of every day. I know because I built it. I designed the Machine to detect acts of terror but it sees everything. Violent crimes involving ordinary people. People like you. Crimes the government considered "irrelevant". They wouldn't act so I decided I would. But I needed a partner. Someone with the skills to intervene. Hunted by the authorities, we work in secret. You'll never find us. But victim or perpetrator, if your number is up, we'll find you.
        </p>
      </div>
    </div>
  </main>
</DocumentFragment>
`;

exports[`<ShowDetails /> When error occured should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="error"
    role="alert"
  >
    Failed to load show data
  </div>
</DocumentFragment>
`;

exports[`<ShowDetails /> When loading show should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="spinner"
    role="status"
  >
    <span
      class="css-1k4jg2i"
    />
  </div>
</DocumentFragment>
`;
